File: mcp_bridge/tools/review.py

"""Review tools: 20-day history + daily full review"""
import sys, os
sys.path.insert(0, r"D:\quicktiny")
from ths_bridge_v3 import _last_trade_date, _get_effective_date, _pg_exec, HAS_PG, log

def handle_review_history(args):
    days = min(int(args.get("days", 20)), 60)
    if HAS_PG:
        rows = _pg_exec(
            "SELECT trade_date, count(*) FILTER (WHERE change_pct >= 9.8) as up_cnt, "
            "count(*) FILTER (WHERE change_pct <= -9.8) as down_cnt, "
            "round(avg(change_pct)::numeric, 2) as avg_pct, sum(amount) as total_amount "
            "FROM daily_kline "
            "WHERE trade_date >= (SELECT max(trade_date) FROM daily_kline) - INTERVAL '%s days' "
            "GROUP BY trade_date ORDER BY trade_date DESC LIMIT %s", (days*2, days))
        if rows:
            lines = [f"=== {days}????? ==="]
            for r in rows:
                d = str(r["trade_date"])[:10] if hasattr(r["trade_date"], "isoformat") else str(r["trade_date"])
                lines.append(f"{d} ??{int(r['up_cnt'] or 0)} ??{int(r['down_cnt'] or 0)} ??{float(r['avg_pct'] or 0):+.2f}%")
            return "\n".join(lines)
    return "???????"

def handle_review_daily(args):
    target = args.get("date", "") or _get_effective_date()
    if HAS_PG and target:
        lines = [f"=== {target} ?????? ==="]
        # emotion
        er = _pg_exec(
            "SELECT count(*) FILTER (WHERE change_pct >= 9.8) as up, "
            "count(*) FILTER (WHERE change_pct <= -9.8) as down, sum(amount) as amt "
            "FROM daily_kline WHERE trade_date=%s", (target,))
        if er:
            lines.append(f"??{er[0]['up'] or 0} ??{er[0]['down'] or 0} ??{(float(er[0]['amt'] or 0)/1e12):.2f}??")
        # ladder
        lr = _pg_exec(
            "SELECT dlu.code, dlu.continue_num, bs.name FROM daily_limit_up dlu "
            "LEFT JOIN base_stocks bs ON dlu.code = bs.code "
            "WHERE dlu.trade_date=%s ORDER BY dlu.continue_num DESC", (target,))
        if lr:
            ladder = {}
            for r in lr:
                lv = r["continue_num"] or 1
                ladder.setdefault(lv, {"level": lv, "names": []})
                ladder[lv]["names"].append(r.get("name", r["code"]))
            for lv in sorted(ladder, reverse=True):
                lines.append(f"{lv}?: {', '.join(ladder[lv]['names'][:10])}")
        return "\n".join(lines)
    return "???????"

REVIEW_TOOLS = {
    "review_history": {
        "description": "[??] ????N?(??20?)?????: ?????/???/???/??????????",
        "handler": handle_review_history,
    },
    "review_daily": {
        "description": "[??] ?????????????: ????+????+???????: date(YYYY-MM-DD)",
        "handler": handle_review_daily,
    },
}

File: supabase/functions/mcp/tools/review.ts
import type { ToolDef } from "./index.ts";

async function lastTradeDate(supabase: any): Promise<string | null> {
  const { data } = await supabase
    .from("daily_kline").select("trade_date").order("trade_date", { ascending: false }).limit(1);
  return data?.[0]?.trade_date ?? null;
}

export const REVIEW_TOOLS: Record<string, ToolDef> = {

  review_history: {
    description: "[??] ?N?(??20?)????: ???/???/???",
    handler: async ({ supabase, args }) => {
      const days = Math.min(Number(args.days) || 20, 60);
      const dt = await lastTradeDate(supabase);
      if (!dt) return "[????] ???";

      const { data } = await supabase
        .from("daily_kline")
        .select("trade_date, change_pct, amount")
        .gte("trade_date", `${dt}::date - interval '${days * 2} days'`)
        .order("trade_date", { ascending: false });

      if (!data?.length) return "[????] ???";

      // Group by trade_date
      const byDate = new Map<string, { up: number; down: number; total: number; sumPct: number; sumAmt: number }>();
      for (const r of data as any[]) {
        const d = r.trade_date;
        if (!byDate.has(d)) byDate.set(d, { up: 0, down: 0, total: 0, sumPct: 0, sumAmt: 0 });
        const g = byDate.get(d)!;
        const pct = Number(r.change_pct ?? 0);
        if (pct >= 9.8) g.up++;
        else if (pct <= -9.8) g.down++;
        g.total++;
        g.sumPct += pct;
        g.sumAmt += Number(r.amount ?? 0);
      }

      const sorted = [...byDate.entries()]
        .sort((a, b) => b[0].localeCompare(a[0]))
        .slice(0, days);

      const lines: string[] = ["=== ???? ==="];
      lines.push("??        | ?? | ?? | ???  | ???");
      for (const [d, g] of sorted) {
        const avgPct = g.total > 0 ? (g.sumPct / g.total).toFixed(2) : "0.00";
        const amt = (g.sumAmt / 1e8).toFixed(0);
        lines.push(`${d} | ${String(g.up).padStart(4)} | ${String(g.down).padStart(4)} | ${avgPct.padStart(6)}% | ${amt}?`);
      }

      return lines.join("\n");
    },
  },

  review_daily: {
    description: "[??] ??????: ????+????+????",
    handler: async ({ supabase, args }) => {
      const date = (args.date as string) || "";
      const dt = date || await lastTradeDate(supabase);
      if (!dt) return "[????] ???";

      const lines: string[] = [`=== ???? (${dt}) ===`];

      // Market overview
      const { count: up } = await supabase
        .from("daily_kline").select("*", { count: "exact", head: true })
        .eq("trade_date", dt).gte("change_pct", 9.8);
      const { count: down } = await supabase
        .from("daily_kline").select("*", { count: "exact", head: true })
        .eq("trade_date", dt).lte("change_pct", -9.8);

      lines.push(`???: ${up ?? 0}? / ${down ?? 0}?`);

      // Top limit-up stocks
      const { data: topUp } = await supabase
        .from("daily_limit_up")
        .select("code, name:base_stocks!inner(name), continue_num, reason_info")
        .eq("trade_date", dt)
        .order("continue_num", { ascending: false })
        .limit(10);

      if (topUp?.length) {
        lines.push("\n???? TOP10:");
        for (const s of topUp as any[]) {
          lines.push(`  ${s.continue_num}? ${s.name?.name ?? s.code}(${s.code}) ${s.reason_info || ""}`);
        }
      }

      // Top themes
      const { data: themes } = await supabase
        .from("daily_limit_up")
        .select("reason_type, code")
        .eq("trade_date", dt);

      if (themes?.length) {
        const themeCount = new Map<string, number>();
        for (const t of themes as any[]) {
          const th = t.reason_type || "??";
          themeCount.set(th, (themeCount.get(th) ?? 0) + 1);
        }
        const sorted = [...themeCount.entries()].sort((a, b) => b[1] - a[1]).slice(0, 10);
        lines.push("\n????:");
        for (const [th, cnt] of sorted) {
          lines.push(`  ${th}: ${cnt}?`);
        }
      }

      return lines.join("\n");
    },
  },
};
